From cd605a3f7f427863b5d5ccc1e545290d36bafe0f Mon Sep 17 00:00:00 2001
From: Ruchika Gupta <ruchika.gupta@freescale.com>
Date: Wed, 24 Apr 2013 10:12:48 +0530
Subject: [PATCH 072/547] crypto: caam - Skip RNG instantiation if RNG already
 instantiated

For SEC with RNG version >= 4, RNG init is performed.
However for secure boot scenarios, there may be a case where
RNG has already been instantiated by u-boot(B4860) or boot ROM
code (C290). In such SoCs, if RNG is instantiated again by crypto
driver, it returns "Instantiation error". RNG4 DRNG STATUS register
has the status bit to indicate if RNG has already been instantiated

Signed-off-by: Ruchika Gupta <ruchika.gupta@freescale.com>
Change-Id: I15d4700f04d18687397c5b8e20461f2457652fad
Reviewed-on: http://git.am.freescale.net:8181/1676
Reviewed-by: Garg Vakul-B16394 <vakul@freescale.com>
Reviewed-by: Geanta Neag Horia Ioan-B05471 <horia.geanta@freescale.com>
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
[Original patch taken from QorIQ-SDK-V1.4-SOURCE-20130625-yocto.iso]
Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 drivers/crypto/caam/ctrl.c | 11 ++++++-----
 drivers/crypto/caam/regs.h |  5 ++++-
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/crypto/caam/ctrl.c b/drivers/crypto/caam/ctrl.c
index a3cbdf1..f5d6dec 100644
--- a/drivers/crypto/caam/ctrl.c
+++ b/drivers/crypto/caam/ctrl.c
@@ -294,13 +294,14 @@ static int caam_probe(struct platform_device *pdev)
 		return -ENOMEM;
 	}
 
-	/*
-	 * RNG4 based SECs need special initialization prior
-	 * to executing any descriptors
-	 */
 	cha_vid = rd_reg64(&topregs->ctrl.perfmon.cha_id);
 
-	if ((cha_vid & CHA_ID_RNG_MASK) >> CHA_ID_RNG_SHIFT >= 4) {
+	/*
+	 * If SEC has RNG version >= 4 and RNG state handle has not been
+	 * already instantiated ,do RNG instantiation
+	 */
+	if ((cha_vid & CHA_ID_RNG_MASK) >> CHA_ID_RNG_SHIFT >= 4 &&
+	    !(rd_reg32(&topregs->ctrl.r4tst[0].rdsta) & RDSTA_IF0)) {
 		kick_trng(pdev);
 		ret = instantiate_rng(ctrlpriv->jrdev[0]);
 		if (ret) {
diff --git a/drivers/crypto/caam/regs.h b/drivers/crypto/caam/regs.h
index eae76f5..c09142f 100644
--- a/drivers/crypto/caam/regs.h
+++ b/drivers/crypto/caam/regs.h
@@ -265,7 +265,10 @@ struct rng4tst {
 		u32 rtfrqmax;	/* PRGM=1: freq. count max. limit register */
 		u32 rtfrqcnt;	/* PRGM=0: freq. count register */
 	};
-	u32 rsvd1[56];
+	u32 rsvd1[40];
+#define RDSTA_IF0 0x00000001
+	u32 rdsta;
+	u32 rsvd2[15];
 };
 
 /*
-- 
1.8.4.93.g57e4c17

